lib/assert.js

```diff
@@ -476,23 +474,44 @@ function partiallyCompareArrayBuffersOrViews(actual, expected) {
 
 function partiallyCompareSets(actual, expected, comparedObjects) {
   if (SetPrototypeGetSize(expected) > SetPrototypeGetSize(actual)) {
-    return false; // `expected` can't be a subset if it has more elements
+    return false;
   }
 
   if (isDeepEqual === undefined) lazyLoadComparison();
 
-  const actualArray = ArrayFrom(FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual));
   const expectedIterator = FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], expected);
-  const usedIndices = new SafeSet();
+  let actualSet;
+
+  for (const expectedItem of expectedIterator) {
+    let foundMatch = false;
+
+    // Check for direct inclusion first (fast path for both primitives and identical object references)
+    if (actual.has(expectedItem)) {
+      foundMatch = true;
+    } else if (!isPrimitive(expectedItem)) {
+      // Only create actualSet if we need it, and only add objects to it
+      if (actualSet === undefined) {
+        actualSet = new SafeSet();
+
+        for (const item of FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual)) {
+          if (!isPrimitive(item)) {
+            actualSet.add(item);
+          }
+        }
```

Comment:
```suggestion
        actualSet.add(item);
```

We already know it's an object and iterating over everything every time again is costly. We can just continue with this loop and move the next loop after the current loop. That way it's also less indentation and hopefully easier to follow.

Reply:
I actually took another approach to more separate data types in order to avoid unnecessary checks and added more specific benchmarks, hopefully it all makes sense 😄